File: Textbook_Exercises/Python/Chapter_9/Exercise_9_1/K_means_demo.py
# This file is associated with the book
# "Machine Learning Refined", Cambridge University Press, 2016.
# by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

import numpy as np
import numpy.matlib
import matplotlib.pyplot as plt
from __future__ import division

def K_means_demo():

    # load data
    X = np.array(np.genfromtxt('Kmeans_demo_data.csv', delimiter=','))

    C0 = np.array([[0,0],[0,.5]])   # initial centroid locations

    # run K-means
    K = np.shape(C0)[1]

    C, W = your_K_means(X, K, C0)

    # plot results
    plot_results(X, C, W, C0)
    plt.show()


def your_K_means(X, K, C0):

# --->  YOUR CODE GOES HERE.

    return C, W

def plot_results(X, C, W, C0):

    K = np.shape(C)[1]

    # plot original data
    fig = plt.figure(facecolor = 'white')
    ax1 = fig.add_subplot(121)
    plt.scatter(X[0,:],X[1,:], s = 50, facecolors = 'k')
    plt.title('original data')
    ax1.set_xlim(-.55, .55)
    ax1.set_ylim(-.55, .55)
    ax1.set_aspect('equal')

    plt.scatter(C0[0,0],C0[1,0],s = 100, marker=(5, 2), facecolors = 'b')
    plt.scatter(C0[0,1],C0[1,1],s = 100, marker=(5, 2), facecolors = 'r')

    # plot clustered data
    ax2 = fig.add_subplot(122)
    colors = ['b','r']

    for k in np.arange(0,K):
        ind = np.nonzero(W[k][:]==1)[0]
        plt.scatter(X[0,ind],X[1,ind],s = 50, facecolors = colors[k])
        plt.scatter(C[0,k],C[1,k], s = 100, marker=(5, 2), facecolors = colors[k])

    plt.title('clustered data')
    ax2.set_xlim(-.55, .55)
    ax2.set_ylim(-.55, .55)
    ax2.set_aspect('equal')

K_means_demo()

File: mlrefined_exercises/ed_1/MATLAB/Chapter_9/Exercise_9_1/K_means_demo.m
function K_means_demo()

% This file is associated with the book
% "Machine Learning Refined", Cambridge University Press, 2016.
% by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

% load data
X = csvread('kmeans_demo_data.csv');

C0 = [0,0;-.5,.5];     % initial centroid locations

% run K-means
K = size(C0,2);
[C, W] = your_K_means(X, K, C0);

% plot results
plot_results(X, C, W, C0);


function [C, W] = your_K_means(X, K, C0)

% ----->  YOUR CODE GOES HERE    
   
end


function plot_results(X, C, W, C0)
    
    K = size(C,2);
    
    % plot original data 
    subplot(1,2,1)
    scatter(X(1,:),X(2,:),'fill','k');
    title('original data')
    axis([0 1 0 1]);
    set(gcf,'color','w');
    set(gca,'XTick',[]);
    set(gca,'YTick',[]);
    axis square
    axis([-.5 .5 -.5 .5]);
    box on
    hold on
    colors = [0 0 1;
              1 0 0;
              0 1 0;
              1 0 1
              1 1 0
              0 1 1];
          
    for k = 1:K  
        scatter(C0(1,k),C0(2,k),100,'x','MarkerFaceColor',colors(k,:),'MarkerEdgeColor',colors(k,:))
        hold on
    end      
          
                 
    % plot clustered data 
    subplot(1,2,2)
    for k = 1:K
        ind = find(W(k,:) == 1);
        scatter(X(1,ind),X(2,ind),'fill','MarkerFaceColor',colors(k,:),'MarkerEdgeColor',colors(k,:));
        hold on
    end
    
    for k = 1:K  
        scatter(C(1,k),C(2,k),100,'x','MarkerFaceColor',colors(k,:),'MarkerEdgeColor',colors(k,:))
        hold on
    end  
    title('clustered data')
    axis([-.5 .5 -.5 .5]);
    set(gcf,'color','w');
    set(gca, 'XTick', []);
    set(gca, 'YTick', []);
    axis square
    box on
end

end



